apps/dgst.c

```diff
@@ -461,7 +461,7 @@ int do_fp(BIO *out, unsigned char *buf, BIO *bp, int sep, int binout,
     size_t len;
     int i, backslash = 0;
 
-    for (;;) {
+    while (!BIO_eof(bp)) {
         i = BIO_read(bp, (char *)buf, BUFSIZE);
```

Comment:
when I look down at line 501, I see something *strange*:
Although bp is clearly at EOF now, this is done:
```C
    if (key != NULL) {
       [...]
    } else {
        len = BIO_gets(bp, (char *)buf, BUFSIZE);
        if ((int)len < 0) {
            ERR_print_errors(bio_err);
            return 1;
        }
    }
```
and, although at EOF the BIO_gets returns 32 bytes (sha signature)
when I do `openssl dgst -sha256`, it is especially unusual, since
gets is supposed to return a zero-terminated string with a new line char
at the end!  The string is neither zero-terminated nor has a new line.
I know at least one place where gets is used in a context that would not
work without zero-termination:
https://github.com/openssl/openssl/blob/38f6f99cdf0a87345d646d30a764c089c38627ad/apps/ocsp.c#L1413-L1420

Reply:
This behaviour of message-digest BIO-filters is documented here: https://www.openssl.org/docs/man1.1.0/man3/BIO_f_md.html
Quote:
_The lack of support for BIO_puts() and the non standard behaviour of BIO_gets() could be regarded as anomalous. It could be argued that BIO_gets() and BIO_puts() should be passed to the next BIO in the chain and digest the data passed through and that digests should be retrieved using a separate BIO_ctrl() call._